cmd/admin-heal.go

```diff
@@ -72,304 +76,556 @@ EXAMPLES:
+    6. Issue a dry-run heal operation to list all objects to be healed under 'dir' prefix
+       $ {{.HelpName}} --recursive --dry-run play/testbucket/dir/
 
-    7. Issue a fake heal operation to list all objects to be healed under 'dir' prefix
-       $ {{.HelpName}} --recursive --fake play/testbucket/dir/
-
-    8. Issue a fake heal operation to list all uploads to be healed under 'dir' prefix
-       $ {{.HelpName}} --incomplete --recursive --fake play/testbucket/dir/
+    7. Force start a heal operation (in case another heal operation is running)
+       $ {{.HelpName}} --recursive --force-start play/testbucket/dir/
 
 `,
 }
 
-// healMessage container to hold repair information.
-type healMessage struct {
-	Status string             `json:"status"`
-	Result madmin.HealResult  `json:"result"`
-	Bucket string             `json:"bucket"`
-	Object *madmin.ObjectInfo `json:"object"`
-	Upload *madmin.UploadInfo `json:"upload"`
-}
+const (
+	lineWidth = 80
+)
+
+// An alias of string to represent the health color code of an object
+type hCol string
 
-// adminHealBefore used to provide users with temporary warning message
-func adminHealBefore(ctx *cli.Context) error {
-	color.Yellow("\t *** mc admin heal is EXPERIMENTAL ***")
-	return setGlobalsFromContext(ctx)
+const (
+	hColGrey   hCol = "Grey"
+	hColRed         = "Red"
+	hColYellow      = "Yellow"
+	hColGreen       = "Green"
+)
+
+// getHPrintCol - map color code to color
+func getHPrintCol(c hCol) *color.Color {
+	switch c {
+	case hColGrey:
+		return color.New(color.FgWhite, color.Bold)
+	case hColRed:
+		return color.New(color.FgRed, color.Bold)
+	case hColYellow:
+		return color.New(color.FgYellow, color.Bold)
+	case hColGreen:
+		return color.New(color.FgGreen, color.Bold)
+	}
+	return nil
 }
 
-// String colorized service status message.
-func (u healMessage) String() string {
-	allOfflineTmpl := "is not healed since many disks were offline"
-	allHealedTmpl := "is healed on all disks"
-	someHealedTmpl := "is healed on some disks while other disks were offline"
+var (
+	hColOrder = []hCol{hColRed, hColYellow, hColGreen}
+	hColTable = map[int][]int{
+		1: {0, -1, 1},
+		2: {0, 1, 2},
+		3: {1, 2, 3},
+		4: {1, 2, 4},
+		5: {1, 3, 5},
+		6: {2, 4, 6},
+		7: {2, 4, 7},
+		8: {2, 5, 8},
+	}
+)
 
-	msg := ""
-	if u.Object != nil {
-		msg = fmt.Sprintf("Object %s/%s ", u.Bucket, u.Object.Key)
-	} else {
-		msg = fmt.Sprintf("Upload %s/%s/%s ", u.Bucket, u.Upload.Key, u.Upload.UploadID)
+func getHColCode(surplusShards, parityShards int) (c hCol, err error) {
+	if parityShards < 1 || parityShards > 8 || surplusShards > parityShards {
+		return c, fmt.Errorf("Invalid parity shard count/surplus shard count given")
+	}
+	if surplusShards < 0 {
+		return hColGrey, err
 	}
+	colRow := hColTable[parityShards]
+	for index, val := range colRow {
+		if val != -1 && surplusShards <= val {
+			return hColOrder[index], err
+		}
+	}
+	return c, fmt.Errorf("this will not be returned")
+}
+
+type hri struct {
```

Comment:
Any reason why we declare this struct ? wouldn't be enough if we use madmin.HealResultItem ?

Reply:
Can you comment this struct ?